returnn/tf/layers/basic.py

```diff
@@ -913,37 +913,199 @@ def transform_config_dict(cls, d, network, get_layer):
     """
-    :param str axis:
-    :param int position:
+    :param LayerBase position: Layer containing the indices used to select the slices of the input from.
+    If another layer, must be of type ``int32`` or ``int64``. Can also specify a constant ``int``.
+    :param str axis: The axis into which we gather the indices into
     """
     super(GatherLayer, self).__init__(**kwargs)
-    axis = self.input_data.get_axis_from_description(axis, allow_int=False)
-    assert isinstance(position, int)  # nothing else implemented yet
-    slices = [slice(None, None)] * axis + [position]  # type: typing.List[typing.Union[slice, int]]
-    self.output.placeholder = self.input_data.placeholder[slices]
+    self.position = position
+
+    input_data = self.input_data
+    position_data = position.output
+    old_gather_axis = input_data.get_axis_from_description(axis, allow_int=False)  # might be moved later
+
+    # determine all common axes of input_data and position_data
+    axes = self._get_common_input_position_axes(input_data, position_data, old_gather_axis)
+    common_axes_input, common_axes_position, input_axes, position_axes = axes
+    batch_dims = len(common_axes_input)
+
+    # move gather axis back if necessary s.t. common axes are always in front
+    gather_axis = max(batch_dims, old_gather_axis)
+
+    # (BatchAxes.., gather_axis, InputAxes..)
+    params = tf.transpose(
+      input_data.placeholder,
+      common_axes_input + input_axes[:gather_axis-batch_dims] + [old_gather_axis] + input_axes[gather_axis-batch_dims:])
+    # (BatchAxes.., PositionAxes..)
+    indices = tf.transpose(position_data.placeholder, common_axes_position + position_axes)
+    # (BatchAxes.., PositionAxes.., InputAxes..)
+    self.output.placeholder = tf.gather(params=params, indices=indices, axis=gather_axis, batch_dims=batch_dims)
+
+    # Take size_placeholder from input_data if they exist there, otherwise from position_data
+    size_placeholder = dict()
+    for input_axis, size in input_data.size_placeholder.items():
+      input_axis = input_data.get_batch_axis(input_axis)
+      if input_axis == old_gather_axis:
+        continue
+      output_axis = self.output.get_batch_axis_excluding_batch(
+        self.translate_input_axis(input_axis, old_gather_axis, *axes))
+      size_placeholder[output_axis] = size
+    for position_axis, size in position_data.size_placeholder.items():
+      position_axis = position_data.get_batch_axis(position_axis)
+      output_axis = self.output.get_batch_axis_excluding_batch(self.translate_position_axis(position_axis, *axes))
+      size_placeholder.setdefault(output_axis, size)
+    self.output.size_placeholder = size_placeholder
+
+  @classmethod
+  def _get_common_input_position_axes(cls, input_data, position_data, old_gather_axis):
+    """
+    Determine all common axes of input_data and position_data. All other axes will be specific axes then (except for
+    the gather axis of ``input_data``)
```

Comment:
```suggestion
    the gather axis of ``input_data``)

```
(Newline after description, before params.)